compiler/grainformat/comment_utils.re

```diff
@@ -0,0 +1,519 @@
+    let (_, cmtline, cmtchar, _) = Locations.get_raw_pos_info(c_loc.loc_end);
+    if (cmtline < line) {
+      [c] @ get_comments_before_line(line, List.tl(comments));
+    } else {
+      []; // can stop early as there will be no more
+    };
+  };
+
+let rec get_comments_after_line = (line: int, comments) =>
+  if (List.length(comments) == 0) {
+    [];
+  } else {
+    let c: Grain_parsing.Parsetree.comment = List.hd(comments);
+    let c_loc: Grain_parsing.Location.t = Locations.get_comment_loc(c);
+    let (_, cmtline, cmtchar, _) = Locations.get_raw_pos_info(c_loc.loc_end);
+    if (cmtline > line) {
+      [c] @ get_comments_after_line(line, List.tl(comments));
+    } else {
+      get_comments_after_line(
+        line,
+        List.tl(comments) // gotta keep checking
+      );
+    };
+  };
+
+let print_comments = (comments: list(Grain_parsing.Parsetree.comment)) => {
+  List.map(c => Comments.print_comment(c), comments);
+};
+
+let rec get_comments_between_lines = (line1: int, line2: int, comments) =>
+  if (List.length(comments) == 0) {
+    [];
+  } else {
+    let c: Grain_parsing.Parsetree.comment = List.hd(comments);
+    let c_loc: Grain_parsing.Location.t = Locations.get_comment_loc(c);
+    let (_, cmtsline, cmtschar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    let (_, cmteline, cmtechar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_end);
+
+    let cmts =
+      if (cmtsline > line1) {
+        if (cmteline < line2) {
+          [c] @ get_comments_between_lines(line1, line2, List.tl(comments));
+        } else
+          {
+            [];
+          }; // can stop early
+      } else {
+        // it's before, so keep going
+        get_comments_between_lines(
+          line1,
+          line2,
+          List.tl(comments),
+        );
+      };
+    cmts;
+  };
+
+let rec get_comments_inside_location =
+        (
+          ~location: Grain_parsing.Location.t,
+          comments: list(Grain_parsing.Parsetree.comment),
+        ) => {
+  let (_, stmtStartLine, stmsStartChar, _) =
+    Locations.get_raw_pos_info(location.loc_start);
+  let (_, stmtEndLine, stmsEndChar, _) =
+    Locations.get_raw_pos_info(location.loc_end);
+
+  if (List.length(comments) > 0) {
+    let cmt: Grain_parsing__Parsetree.comment = List.hd(comments);
+    let c_loc: Grain_parsing.Location.t = Locations.get_comment_loc(cmt);
+
+    let (_, cmtsline, cmtschar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    let (_, cmteline, cmtechar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_end);
+    if (cmtsline > stmtEndLine) {
+      []; // can stop now
+    } else if (cmteline < stmtStartLine) {
+      get_comments_inside_location(location, List.tl(comments));
+    } else if
+      // need the complex comparisons here
+      (cmtsline > stmtStartLine
+       || cmtsline == stmtStartLine
+       && cmtschar >= stmsStartChar) {
+      if (cmteline < stmtEndLine
+          || cmteline == stmtEndLine
```

Comment:
```suggestion
      if (cmteline <= stmtEndLine
```

Reply:
You know, I'm not sure.  The logic  I want is if cmteline < stmtEndLine I don't care about cmtechar, but if cmteline = stmtEndLine I do.   As the formatter took away my parens I think that's what the operator priority gives with the following && but I need to work that out.

Reply:
@ospencer  I tested and I think this is right, because the && after is dependent on the first part of the or clause

In English, if the cmt ends on a before  the statement we don't take the end char into account, only if the line is the same do we (because the end char on a line before might be after the start char on a line after) if that makes sense